### PR TITLE
Refactor native modules into separate modules

### PR DESCRIPTION
**Pull request by siddarthkay**
### Summary

The native module : `react-native-status` has always been 1 big fat file per platform which contained all of the native modules.

In case of Android it was `StatusModule.java`
This PR modularises it & attempts to thin out this 1 big file into these categories:  
- `AccountManager.java`
- `EncryptionUtils.java`
- `DatabaseManager.java`
- `UIHelper.java`
- `LogManager.java` 
- `NetworkManager.java` 
- `Utils.java`

In case of iOS it was `RCTStatus.m`
This PR modularises it & attempts to thin out this 1 big file into these categories:  
- `AccountManager.m`
- `EncryptionUtils.m`
- `DatabaseManager.m`
- `UIHelper.m`
- `LogManager.m` 
- `NetworkManager.m` 
- `Utils.m`

In this PR we also remove a lot of unused native code which has no reference on cljs side. 

### Testing notes
In theory there should be no change in functionality, but since code has been moved around which talks to status-go , please do a thorough check of this PR.

#### Platforms
- Android
- iOS

status: ready

### CODE REVIEW

File: src/legacy/status_im/utils/test.cljs [hunk 1]
```diff
@@ -31,108 +31,100 @@
...
+                 deserialization-key)))}))
 
-    :callPrivateRPC
-    (fn [payload callback]
-      (callback (.callPrivateRPC native-status payload)))
-
-    :createAccountAndLogin (fn [request] (.createAccountAndLogin native-status request))
-
-    :saveAccountAndLogin
+(def account-manager
```
Comment (ilmotta) [inline src/legacy/status_im/utils/test.cljs:66]: I don't know about you and other folks, but the lack of consistent whitespace in the vars `account-manager`, `utils`, and `encryption-utils` makes reading them uncomfortable.

My feedback about the code style is to separate different keys and values by a line break, and to consistently move the function body below the arguments' vector in each `(fn [...] <function body>)`. This makes scanning the code vertically a breeze.

```clojure
(def account-manager
  (clj->js
   {:saveAccountAndLogin
    (fn [multiaccount-data password settings config accounts-data]
      (.saveAccountAndLogin native-status multiaccount-data password settings config accounts-data))

    :openAccounts
    (fn [callback]
      (callback (.openAccounts native-status test-dir)))

    :createAccountAndLogin
    (fn [request]
      (.createAccountAndLogin native-status request))

    :logout
    (fn []
      (.logout native-status))

    :multiAccountImportMnemonic
    (fn [json callback]
      (callback (.multiAccountImportMnemonic native-status json)))

    :multiAccountLoadAccount
    (fn [json callback]
      (callback (.multiAccountLoadAccount native-status json)))

    :multiAccountDeriveAddresses
    (fn [json callback]
      (callback (.multiAccountDeriveAddresses native-status json)))

    :multiAccountGenerateAndDeriveAddresses
    (fn [json callback]
      (callback (.multiAccountGenerateAndDeriveAddresses native-status json)))

    :multiAccountStoreDerived
    (fn [json callback]
      (callback (.multiAccountStoreDerivedAccounts native-status json)))}))
```

File: src/native_module/core.cljs [hunk 2]
```diff
@@ -11,6 +11,41 @@
   (when (exists? (.-NativeModules react-native))
     (.-Status ^js (.-NativeModules react-native))))
 
+(defn account-manager
+  []
+  (when (exists? (.-NativeModules react-native))
+    (.-AccountManager ^js (.-NativeModules react-native))))
```
Comment (ilmotta) [inline src/native_module/core.cljs:17]: Why do we need to make `account-manager` and its friends functions? Couldn't we define them as static values? It seems we just grab static properties from `react-native`.

Or is this a performance trick to reduce memory consumption at startup?

File: modules/react-native-status/ios/RCTStatus/NetworkManager.m [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+RCT_EXPORT_METHOD(recover:(NSString *)message
+        callback:(RCTResponseSenderBlock)callback) {
+#if DEBUG
+    NSLog(@"Recover() method called");
+#endif
+    NSString *result = StatusgoRecover(message);
+    callback(@[result]);
+}
+
+@end
```
Comment (ilmotta) [inline modules/react-native-status/ios/RCTStatus/NetworkManager.m]: Missing trailing newline character here and in a few other files.

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/LogManager.java [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        jsonConfig.put("File", getLogsFile().getAbsolutePath());
+        final String config = jsonConfig.toString();
+        this.utils.executeRunnableStatusGoMethod(() -> Statusgo.initLogging(config), callback);
+    }
+
+    @ReactMethod(isBlockingSynchronousMethod = true)
+    public String logFileDirectory() {
+        return this.utils.getPublicStorageDirectory().getAbsolutePath();
+    }
+
```
Comment (ilmotta) [inline modules/react-native-status/android/src/main/java/im/status/ethereum/module/LogManager.java:226]: Extra line.

**Review by ilmotta — APPROVED** [review]
Fabulous work @siddarthkay 🥂 

I quickly scanned all files, but it's too much work to compare the green code with the red one from the god modules. In Sid we trust.

**Review by mohsen-ghafouri — APPROVED** [review]
Amazing work 🙂

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/AccountManager.java [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+    }
+
+    @ReactMethod
+    public void prepareDirAndUpdateConfig(final String keyUID, final String config, final Callback callback) {
+        Log.d(TAG, "prepareDirAndUpdateConfig");
+        String finalConfig = prepareDirAndUpdateConfig(config, keyUID);
+        callback.invoke(finalConfig);
+    }
+
+    //TODO : maybe nuke since it is not called anywhere in status-mobile code
```
Comment (siddarthkay) [inline modules/react-native-status/android/src/main/java/im/status/ethereum/module/AccountManager.java]: I missed nuking these, will take care of these

File: modules/react-native-status/ios/RCTStatus/UIHelper.h [hunk 6]
```diff
@@ -0,0 +1,9 @@
+#import <sys/utsname.h>
+#import <Foundation/Foundation.h>
+#import <React/RCTBridgeModule.h>
+#import "Statusgo.h"
```
Comment (siddarthkay) [inline modules/react-native-status/ios/RCTStatus/UIHelper.h]: safe to remove this import from here

File: src/legacy/status_im/utils/test.cljs (hunk 1 above)
Comment (siddarthkay) [reply src/legacy/status_im/utils/test.cljs:66]: Ah yes thanks for this suggestion, I did keep a mental note back when I was fixing tests that this formatting did look weird.
I later forgot about it and assumed that `make lint-fix` would tidy things up 😄 
I'll tidy this up.

File: src/native_module/core.cljs (hunk 2 above)
Comment (siddarthkay) [reply src/native_module/core.cljs:17]: I'm not really sure of the perf impacts of making it a `defn` vs `def`.

To be honest I simply followed the existing pattern and the way things were accessed was by grabbing these properties from `NativeModules` and storing them in functions.

I did try your suggestion of making them static and found no issues so I guess we can keep them as static for now and update in a separate PR if we find any performance improvements for moving back to functions.

addressed here : https://github.com/status-im/status-mobile/pull/18385/commits/4fdb0d80235a6a519e86f978c6a4d1ed50dc85a5

File: modules/react-native-status/ios/RCTStatus/NetworkManager.m (hunk 3 above)
Comment (siddarthkay) [reply modules/react-native-status/ios/RCTStatus/NetworkManager.m]: fixed and updated script to catch Objective-C files here : https://github.com/status-im/status-mobile/pull/18385/commits/5852012ba9c5e77ca8b9a85d5e36d7c216aeb31e

File: src/legacy/status_im/utils/test.cljs (hunk 1 above)
Comment (siddarthkay) [reply src/legacy/status_im/utils/test.cljs:66]: addressed here : https://github.com/status-im/status-mobile/pull/18385/commits/c877bd3c36c7caa0e99757c9715148cc7e51fa73

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/AccountManager.java (hunk 5 above)
Comment (siddarthkay) [reply modules/react-native-status/android/src/main/java/im/status/ethereum/module/AccountManager.java]: cleaned up here : https://github.com/status-im/status-mobile/pull/18385/commits/1cd73b0242def75538da0e0b3eee2ee18304e6ca

File: modules/react-native-status/ios/RCTStatus/UIHelper.h (hunk 6 above)
Comment (siddarthkay) [reply modules/react-native-status/ios/RCTStatus/UIHelper.h]: cleaned up here : https://github.com/status-im/status-mobile/pull/18385/commits/1cd73b0242def75538da0e0b3eee2ee18304e6ca

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/LogManager.java (hunk 4 above)
Comment (siddarthkay) [reply modules/react-native-status/android/src/main/java/im/status/ethereum/module/LogManager.java:226]: cleaned up here : https://github.com/status-im/status-mobile/pull/18385/commits/1cd73b0242def75538da0e0b3eee2ee18304e6ca

**Comment by siddarthkay** [comment]
> In Sid we trust.

Thanks for your vote of confidence.
Incase I missed something I am sure E2E will catch 🤞🏻

**Comment by siddarthkay** [comment]
Hmm so 0% E2E passed because release `apk` fails with this error screenshot below.
it is worth noting that the debug `apk` works well on local `android` simulator.

android-fail

File: src/native_module/core.cljs (hunk 2 above)
Comment (siddarthkay) [reply src/native_module/core.cljs:17]: @ilmotta : Unfortunately this approach causes a crash in release `apk` , hence I had to revert back to using functions here https://github.com/status-im/status-mobile/pull/18385/commits/2086ec70474299eb5bc5017d9558ad6b30e33e03

**Review by briansztamfater — APPROVED** [review]
🔥

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/NetworkManager.java
Comment (qfrank) [inline modules/react-native-status/android/src/main/java/im/status/ethereum/module/NetworkManager.java:1]: Putting syncing into the NetworkManager source code file makes me feel like we are downplaying the concept of syncing. This is my feeling but not serious. 😅

**Review by qfrank — APPROVED** [review]

**Review by OmarBasem — APPROVED** [review]

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/AccountManager.java [hunk 7]
```diff
@@ -0,0 +1,345 @@
+package im.status.ethereum.module;
```
Comment (OmarBasem) [inline modules/react-native-status/android/src/main/java/im/status/ethereum/module/AccountManager.java:1]: I was wondering if these files can be grouped into sub-directories

Comment (siddarthkay) [reply modules/react-native-status/android/src/main/java/im/status/ethereum/module/AccountManager.java:1]: Indeed they could be.

I was thinking we would have to update directory structure from : 
`modules/react-native-status/android/src/main/java/im/status/ethereum/module/...` to
`modules/react-native-status/android/src/main/kotlin/im/status/ethereum/module/...` when https://github.com/status-im/status-mobile/issues/18310 is complete.

We can then have a discussion on the folder structure and when we reach a consensus on folder structure we can then open a new PR for this.

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/NetworkManager.java
Comment (siddarthkay) [reply modules/react-native-status/android/src/main/java/im/status/ethereum/module/NetworkManager.java:1]: Haha no hard feelings against syncing, at the time of grouping functions it seemed like a good to fit to belong in NetworkManager, however we can always move it to a separate SyncingManager module later 😉

File: src/native_module/core.cljs (hunk 2 above)
Comment (ilmotta) [reply src/native_module/core.cljs:17]: Quite interesting @siddarthkay. Thanks for attempting this refactor. I'm curious about the error, but let's not waste time on that right?

Comment (siddarthkay) [reply src/native_module/core.cljs:17]: It was a cryptic message about unable to load modules on `undefined` I found it via logcat which helped me figure out that the root cause was this commit :)

**Comment by pavloburykh** [comment]
@siddarthkay thanx for the PR. Please have a look at the issue. 

### ISSUE 1 "Pj().deleteMultiaccount is not a function" occurs when deleteing account on IOS

Reproduced on IOS only. (IOS 16.7.2 iPhone X)

**Steps:** 

1. Open Profiles list
2. Try to remove profile
3. Observe the error

**Actual result:** "Pj().deleteMultiaccount is not a function"  error. Cannot delete the profile. 

[status_logs (5).zip](https://github.com/status-im/status-mobile/files/13919563/status_logs.5.zip)

photo_2024-01-12 15 28 47

https://github.com/status-im/status-mobile/assets/97245802/de948e06-d505-454a-9304-7a15fb6d9201



**Expected result:** profile has been removed

**Comment by siddarthkay** [comment]
@pavloburykh : ISSUE 1 should be fixed now.

**Comment by pavloburykh** [comment]
@siddarthkay thank you for the PR and fix. Ready for merge!

**Pull request merged**